pkg/logql/log/filter.go

```diff
@@ -161,16 +164,102 @@ func (r regexpFilter) ToStage() Stage {
 	}
 }
 
+var (
+	// BytesBufferPool is a bytes buffer used for lines decompressed.
+	// Buckets [0.5KB,1KB,2KB,4KB,8KB]
+	BytesBufferPool = pool.New(1<<9, 1<<13, 2, func(size int) interface{} { return make([]byte, 0, size) })
+
+	toLower = func(r rune) rune { return unicode.To(unicode.LowerCase, r) }
+)
+
 type containsFilter struct {
 	match           []byte
 	caseInsensitive bool
+
+	buf []byte // reusable buffer for lowercase transformation
+}
+
+func (l *containsFilter) Filter(line []byte) bool {
+	if !l.caseInsensitive {
+		return bytes.Contains(line, l.match)
+	}
+	// verify if we have uppercase in the line and if it's only ascii chars
+	isASCII, hasUpper := true, false
+	for i := 0; i < len(line); i++ {
+		c := line[i]
+		if c >= utf8.RuneSelf {
+			isASCII = false
+			break
+		}
+		hasUpper = hasUpper || ('A' <= c && c <= 'Z')
+	}
+	if isASCII {
+		if !hasUpper {
+			return bytes.Contains(line, l.match)
+		}
+		return bytes.Contains(l.toLowerASCII(line), l.match)
+	}
+	return bytes.Contains(l.toLowerUnicode(line), l.match)
 }
 
-func (l containsFilter) Filter(line []byte) bool {
-	if l.caseInsensitive {
-		line = bytes.ToLower(line)
+func (l *containsFilter) toLowerASCII(line []byte) []byte {
+	if len(line) > cap(l.buf) {
+		if l.buf != nil {
+			BytesBufferPool.Put(l.buf)
+		}
+		l.buf = BytesBufferPool.Get(len(line)).([]byte)[:len(line)]
+	}
+	for i := 0; i < len(line); i++ {
```

Comment:
If the buffer is being reused, don't we need to set `l.buf l.buf[:len(line)]`? Otherwise, it looks like it could include the end of a previous line that was longer. Alternatively, we could also return `l.buf[:len(line)]` at the end.